### PR TITLE
Fix up fake resolve roots for proc macros

### LINKED ISSUE #2777: Dependency vendoring broken in 0.49.1
0.49.0 included https://github.com/bazelbuild/rules_rust/pull/2749 which introduced the error

```
INFO: Running command line: bazel-bin/third-party/rust/crates_vendor.sh

error: failed to parse manifest at `/tmp/.tmpqANJaE/Cargo.toml`

Caused by:
  this virtual manifest specifies a `dependencies` section, which is not allowed

Error: Failed to run cargo tree: exit status: 101
```

this has, seemingly, been fixed in 0.49.1 by https://github.com/bazelbuild/rules_rust/pull/2775, but a new error is introduced instead:

```
INFO: Running command line: bazel-bin/third-party/rust/crates_vendor.sh

    Updating crates.io index
error: no matching package named `cognite_async_closure` found
location searched: registry `crates-io`
required by package `rules_rust_fake_proc_macro_root v0.0.0 (/tmp/.tmpXwBHRP)`

Error: Failed to run cargo tree: exit status: 101
```

where `cognite_async_closure` is an internal crate defined in the same repository.

### LINKED ISSUE #2783: custom proc macro crate not found in a cargo workspace setup
In a cargo setup, when a proc macro crate is present in a cargo workspace, and one of the crates in the workspace depends on the proc macro crate, dependencies fail to resolve with the following error:

```
Error in fail: /home/developer/.cache/bazel/_bazel_developer/2927fffaf2126be7f97bf2508c6daa6e/modextwd/rules_rust~~crate/cargo-bazel splice --output-dir /home/developer/.cache/bazel/_bazel_developer/2927fffaf2126be7f97bf2508c6daa6e/modextwd/rules_rust~~crate/crates/splicing-output --config /home/developer/.cache/bazel/_bazel_developer/2927fffaf2126be7f97bf2508c6daa6e/modextwd/rules_rust~~crate/crates/config.json --splicing-manifest /home/developer/.cache/bazel/_bazel_developer/2927fffaf2126be7f97bf2508c6daa6e/modextwd/rules_rust~~crate/crates/splicing_manifest.json --cargo-lockfile /workspace/rust/Cargo.lock --cargo /home/developer/.cache/bazel/_bazel_developer/2927fffaf2126be7f97bf2508c6daa6e/external/rules_rust~~rust_host_tools~rust_host_tools/bin/cargo --rustc /home/developer/.cache/bazel/_bazel_developer/2927fffaf2126be7f97bf2508c6daa6e/external/rules_rust~~rust_host_tools~rust_host_tools/bin/rustc returned with exit code 1:

    Blocking waiting for file lock on package cache
    Updating crates.io index
error: no matching package named <the proc macro crate's name> found
location searched: registry `crates-io`
required by package `rules_rust_fake_proc_macro_root v0.0.0 (/tmp/.tmp79Nulh)`

Error: Failed to generate features

Caused by:
    Failed to run cargo tree: exit status: 101
```

this error seems to happen in v0.49.1. everything works fine in v0.48.0

### PR DESCRIPTION
**Pull request by illicitonion**
1. Skip in-workspace proc-macros - they shouldn't be resolved, and are already resolve roots
2. Properly parse out git deps, rather than assuming everything is a registry dep

Fixes #2777
Fixes #2783

### CODE REVIEW

File: crate_universe/src/metadata.rs [hunk 1]
```diff
@@ -715,7 +716,22 @@ impl TreeResolver {
...
-            .map(|pm| (pm.name.clone(), pm.version.to_string()))
+            // Filter out any in-workspace proc macros, populate dependency details for non-in-workspace proc macros.
+            .filter_map(|pm| {
+                if let Some(source) = pm.source.as_ref() {
+                    let mut detail = DependencyDetailWithOrd(cargo_toml::DependencyDetail {
+                        package: Some(pm.name.clone()),
+                        ..cargo_toml::DependencyDetail::default()
+                    });
+
+                    let source = Source::parse(&source.repr, pm.version.to_string()).expect("TODO");
```
Comment (UebelAndre) [inline crate_universe/src/metadata.rs:727]: TODO?

**Review by UebelAndre — CHANGES_REQUESTED** [review]
One nit and do you know if we're regression sufficient now? How did existing coverage miss the two issues linked?

**Comment by illicitonion** [comment]
> How did existing coverage miss the two issues linked?

We previously didn't have any tests/examples where a Cargo workspace contained a first-party proc-macro. We now have this, in the musl_cross_compiling example (`local_proc_macro`). Both of the link issues had this same root cause, which is not filtering those out from the list.

**Review by UebelAndre — APPROVED** [review]
Thanks! Wanna start a `0.49.2` release?

**Comment by illicitonion** [comment]
Will do on merge! Thanks!

**Pull request merged**